### PR TITLE
Mvv widget auto reload

### PR DESCRIPTION
**Pull request by sebmarste**
## Issue
added mode where you have to press the reload button to get new data (default)
added auto mode where the data gets updated automatically (can be selected in configure activity)

### switch to select reload mode
screenshot_20170608-143425
### reload button in widget
screenshot_20170608-143438

### CODE REVIEW

File: app/src/main/java/de/tum/in/tumcampusapp/widgets/MVVWidgetService.java [hunk 1]
```diff
@@ -27,10 +27,12 @@ public RemoteViewsFactory onGetViewFactory(Intent intent) {
         private final Context applicationContext;
         private List<TransportManager.Departure> departures = new ArrayList<>();
         private int appWidgetID;
+        private boolean forceLoadDepartures = false;
 
         MVVRemoteViewFactory(Context applicationContext, Intent intent) {
             this.applicationContext = applicationContext.getApplicationContext();
             this.appWidgetID = intent.getIntExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, -1);
+            this.forceLoadDepartures = intent.getBooleanExtra(MVVWidget.MVV_WIDGET_FORCE_RELOAD, false);
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/widgets/MVVWidgetService.java]: As I do not know this: when is this contructor called? Does it rerun once the setting has been updated?

Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/widgets/MVVWidgetService.java]: Its done by the broadcast above right?

Comment (sebmarste) [inline app/src/main/java/de/tum/in/tumcampusapp/widgets/MVVWidgetService.java]: the Factory gets instantiated by the intend in MVVWidgets updateWidget method. So every time the widged is updated a new factory is used

File: app/src/main/java/de/tum/in/tumcampusapp/managers/TransportManager.java [hunk 2]
```diff
@@ -126,10 +126,10 @@ public TransportManager(Context context) {
                 "id INTEGER PRIMARY KEY AUTOINCREMENT, symbol VARCHAR)");
 
         db.execSQL("CREATE TABLE IF NOT EXISTS transport_widgets (" +
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/managers/TransportManager.java]: A better name would be `widgets_transport`. (Optional)

File: app/src/main/java/de/tum/in/tumcampusapp/managers/TransportManager.java [hunk 3]
```diff
@@ -196,26 +193,21 @@ public void deleteWidget(int widget_id) {
...
-            if (use_location) {
-                // TODO implement nearest station (replace the station_id string with the calculated station)
-                station = "use location";
-            }
         }
-        WidgetDepartures wD = new WidgetDepartures(station, station_id);
-        TransportManager.widgetDepartures.put(widget_id, wD);
-        return wD;
+        TransportManager.widgetDeparturesList.put(widget_id, widgetDepartures);
+        return widgetDepartures;
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/managers/TransportManager.java]: Thanks for properly naming these variables.

File: app/src/main/java/de/tum/in/tumcampusapp/managers/TransportManager.java [hunk 4]
```diff
@@ -407,20 +399,42 @@ public StationResult(String station, String id, int quality) {
...
-        String station;
-        String station_id;
-        List<Departure> departures;
-        long last_load;
-        boolean is_offline = false;
+
+        private String station;
+        private String station_id;
+        private boolean use_location;
+        private boolean auto_reload;
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/managers/TransportManager.java]: Please use camelCase for all variables ;-)

File: app/src/main/java/de/tum/in/tumcampusapp/managers/TransportManager.java [hunk 5]
```diff
@@ -407,20 +399,42 @@ public StationResult(String station, String id, int quality) {
     }
 
     public static class WidgetDepartures {
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/managers/TransportManager.java]: Can you please put this into a separate file? Nested classes are nasty.

File: app/src/main/java/de/tum/in/tumcampusapp/widgets/MVVWidget.java [hunk 6]
```diff
@@ -121,33 +125,36 @@ static void updateAppWidgets(Context context, AppWidgetManager appWidgetManager,
...
+        rv.setTextViewText(R.id.mvv_widget_station, widgetDepartures.getStation());
 
         // Set up offline symbol (may be shown one update delayed)
         rv.setViewVisibility(R.id.mvv_widget_offline, widgetDepartures.isOffline() ? View.VISIBLE : View.INVISIBLE);
 
         // Set up the configuration activity listeners
         Intent configIntent = new Intent(context, MVVWidgetConfigureActivity.class);
         configIntent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId);
-        PendingIntent pendingIntent = PendingIntent.getActivity(context, appWidgetId, configIntent, PendingIntent.FLAG_UPDATE_CURRENT);
+        PendingIntent pendingIntent = PendingIntent.getActivity(context, appWidgetId * 10, configIntent, PendingIntent.FLAG_UPDATE_CURRENT);
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/widgets/MVVWidget.java]: Random constant somewhere in the code? BAAAAD :dagger:

File: app/src/main/java/de/tum/in/tumcampusapp/widgets/MVVWidget.java [hunk 7]
```diff
@@ -121,33 +125,36 @@ static void updateAppWidgets(Context context, AppWidgetManager appWidgetManager,
...
         rv.setOnClickPendingIntent(R.id.mvv_widget_header, pendingIntent);
 
+        // Set up the reload functionality
+        Intent reloadIntent = new Intent(context, MVVWidget.class);
+        reloadIntent.setAction(MVVWidget.MVV_WIDGET_FORCE_RELOAD);
+        reloadIntent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId);
+        PendingIntent pendingReloadIntent = PendingIntent.getBroadcast(context, appWidgetId + 10 + 1, reloadIntent, PendingIntent.FLAG_UPDATE_CURRENT);
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/widgets/MVVWidget.java]: Again. Why is it 10+1 now?

File: build.gradle [hunk 8]
```diff
@@ -15,6 +15,8 @@ buildscript {
 
 allprojects {
     tasks.withType(JavaCompile) {
+        options.fork = true  // Fork your compilation into a child process
+        options.forkOptions.setMemoryMaximumSize("4g") // Set maximum memory to 4g
```
Comment (kordianbruck) [inline build.gradle]: Thanks, probably a good idea to have this on by default.

File: app/src/main/java/de/tum/in/tumcampusapp/widgets/MVVWidget.java (hunk 6 above)
Comment (sebmarste) [inline app/src/main/java/de/tum/in/tumcampusapp/widgets/MVVWidget.java]: well actually this is an id for the intent, so it does not work, like expected, if there are some with the same id
the most simpel solution was to append an intend index to the widget id:
appWidgetID * 10       -> xxx0
appWidgetID * 10 + 1 -> xxx1

Comment (sebmarste) [inline app/src/main/java/de/tum/in/tumcampusapp/widgets/MVVWidget.java]: bin indeed it is not nice, any idea?

File: build.gradle (hunk 8 above)
Comment (sebmarste) [inline build.gradle]: oh, i did not actively push this, but yea it does make sense...

File: app/src/main/java/de/tum/in/tumcampusapp/widgets/MVVWidget.java (hunk 6 above)
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/widgets/MVVWidget.java]: What do you mean by "if there are some with the same id"?

File: app/src/main/java/de/tum/in/tumcampusapp/managers/TransportManager.java (hunk 2 above)
Comment (sebmarste) [inline app/src/main/java/de/tum/in/tumcampusapp/managers/TransportManager.java]: renamed 👍

File: app/src/main/java/de/tum/in/tumcampusapp/managers/TransportManager.java (hunk 5 above)
Comment (sebmarste) [inline app/src/main/java/de/tum/in/tumcampusapp/managers/TransportManager.java]: ok i extracted all the nested classes from the transport manager to a model package called efa

File: app/src/main/java/de/tum/in/tumcampusapp/managers/TransportManager.java (hunk 4 above)
Comment (sebmarste) [inline app/src/main/java/de/tum/in/tumcampusapp/managers/TransportManager.java]: im using camel case now

File: app/src/main/java/de/tum/in/tumcampusapp/widgets/MVVWidget.java (hunk 6 above)
Comment (sebmarste) [inline app/src/main/java/de/tum/in/tumcampusapp/widgets/MVVWidget.java]: the pending intent to open the config activity and the one to reload, usually u take the appWidgetId there, but we have two intents

Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/widgets/MVVWidget.java]: You should use the same id and add a extra/bundle to pass the sub-action.

**Comment by kordianbruck** [comment]
@sebmarste need to fix the ID thing here and resolve the string conflicts, then we can merge this as well

**Comment by sebmarste** [comment]
shure, ill do this right after ive eaten 👍

File: app/src/main/java/de/tum/in/tumcampusapp/widgets/MVVWidget.java (hunk 6 above)
Comment (sebmarste) [inline app/src/main/java/de/tum/in/tumcampusapp/widgets/MVVWidget.java]: seems like the request code does not even make any difference

**Comment by sebmarste** [comment]
i think its ready to merge

**Pull request merged**